lite2/verifier.go

```diff
@@ -16,101 +16,150 @@ var (
+func VerifyAdjacent(
+	chainID string,
+	trustedHeader *types.SignedHeader,
+	untrustedHeader *types.SignedHeader,
+	untrustedVals *types.ValidatorSet,
+	trustingPeriod time.Duration,
+	now time.Time) error {
+
+	if untrustedHeader.Height != trustedHeader.Height+1 {
+		return errors.New("headers must be adjacent in height")
 	}
 
-	if err := verifyNewHeaderAndVals(chainID, h2, h2Vals, h1, now); err != nil {
+	err := initialHeaderVerification(untrustedHeader, trustedHeader, trustingPeriod, now, chainID, untrustedVals)
+	if err != nil {
 		return err
 	}
 
-	if h2.Height == h1.Height+1 {
-		if !bytes.Equal(h2.ValidatorsHash, h1NextVals.Hash()) {
-			err := errors.Errorf("expected old header next validators (%X) to match those from new header (%X)",
-				h1NextVals.Hash(),
-				h2.ValidatorsHash,
-			)
-			return err
-		}
+	// Check the validator hashes are the same
+	if !bytes.Equal(untrustedHeader.ValidatorsHash, trustedHeader.NextValidatorsHash) {
+		err := errors.Errorf("expected old header next validators (%X) to match those from new header (%X)",
+			trustedHeader.NextValidatorsHash,
+			untrustedHeader.ValidatorsHash,
+		)
+		return err
+	}
+	return nil
+}
+
+func Verify(
+	chainID string,
+	trustedHeader *types.SignedHeader,
+	trustedNextVals *types.ValidatorSet,
+	untrustedHeader *types.SignedHeader,
+	untrustedVals *types.ValidatorSet,
+	trustingPeriod time.Duration,
+	now time.Time,
+	trustLevel tmmath.Fraction) error {
+
+	if untrustedHeader.Height != trustedHeader.Height+1 {
+		return VerifyNonAdjacent(chainID, trustedHeader, trustedNextVals, untrustedHeader, untrustedVals,
+			trustingPeriod, now, trustLevel)
 	} else {
-		// Ensure that +`trustLevel` (default 1/3) or more of last trusted validators signed correctly.
-		err := h1NextVals.VerifyCommitTrusting(chainID, h2.Commit.BlockID, h2.Height, h2.Commit, trustLevel)
-		if err != nil {
-			switch e := err.(type) {
-			case types.ErrNotEnoughVotingPowerSigned:
-				return ErrNewValSetCantBeTrusted{e}
-			default:
-				return e
-			}
-		}
+		return VerifyAdjacent(chainID, trustedHeader, untrustedHeader, untrustedVals, trustingPeriod, now)
 	}
 
-	// Ensure that +2/3 of new validators signed correctly.
-	err := h2Vals.VerifyCommit(chainID, h2.Commit.BlockID, h2.Height, h2.Commit)
-	if err != nil {
+}
+
+func initialHeaderVerification(untrustedHeader *types.SignedHeader, trustedHeader *types.SignedHeader,
+	trustingPeriod time.Duration, now time.Time, chainID string, untrustedVals *types.ValidatorSet) error {
+	// Ensure last header can still be trusted.
+	if HeaderExpired(trustedHeader, trustingPeriod, now) {
+		return ErrOldHeaderExpired{trustedHeader.Time.Add(trustingPeriod), now}
+	}
+
+	if err := verifyNewHeaderAndVals(chainID, untrustedHeader, untrustedVals, trustedHeader, now); err != nil {
 		return err
 	}
 
+	// Ensure that +2/3 of new validators signed correctly.
+	if err := untrustedVals.VerifyCommit(chainID, untrustedHeader.Commit.BlockID, untrustedHeader.Height,
```

Comment:
this should really go last - it's quite expensive - you can easily DDOS the light client

Reply:
Okay I tried to combine the common functions called by both verifydajacent and verifynonadjacent, but I can separate them out again